src/v/datalake/coordinator/state_update.cc

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#include "datalake/coordinator/state_update.h"
+
+#include "datalake/coordinator/state.h"
+#include "datalake/coordinator/translated_offset_range.h"
+
+#include <iterator>
+
+namespace datalake::coordinator {
+
+checked<add_files_update, stm_update_error> add_files_update::build(
+  const topics_state& state,
+  const model::topic_partition& tp,
+  chunked_vector<translated_offset_range> entries) {
+    add_files_update update{
+      .tp = tp,
+      .entries = std::move(entries),
+    };
+    if (!update.can_apply(state)) {
+        return stm_update_error::failed;
+    }
+    return update;
+}
+
+bool add_files_update::can_apply(const topics_state& state) {
+    if (entries.empty()) {
+        return false;
+    }
+    auto prt_state_opt = state.partition_state(tp);
+    if (!prt_state_opt.has_value()) {
+        // No entries at all, this partition hasn't ever added any files.
+        return true;
+    }
+    const auto& prt_state = prt_state_opt.value().get();
+    if (
+      prt_state.pending_entries.empty()
+      && !prt_state.last_committed.has_value()) {
+        // No entries at all, this partition hasn't ever added any files.
+        return true;
+    }
+    auto last_added_offset
+      = prt_state.pending_entries.empty()
+          ? prt_state.last_committed.value()
+          : kafka::offset{prt_state.pending_entries.back().last_offset()};
+    auto update_range_start_offset = entries.begin()->start_offset;
+    if (kafka::next_offset(last_added_offset) == update_range_start_offset) {
```

Comment:
nit: we also have `model::next_offset` so we do not have to depend on kafka here

Reply:
Using kafka offsets here is actually intentional since the state we're managing is the user data. I'm thinking about the coordinator like a Kafka application.